src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+    val unmatchedUmi = "N" * umiLength
+    val umiMetrics   = (umiSequences ++ Seq(unmatchedUmi)).map(umi => umi -> UmiCorrectionMetrics(umi)).toMap
+
+    // Now go through and correct the UMIs in the BAM
+    var (totalRecords, missingUmisRecords, wrongLengthRecords, mismatchedRecords) = (0L, 0L, 0L, 0L)
+    val in        = SamReaderFactory.make().open(input)
+    val out       = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(in.getFileHeader, true, output.toFile, null)
+    val rejectOut = rejects.map(r => new SAMFileWriterFactory().setCreateIndex(true).makeWriter(in.getFileHeader, true, r.toFile, null))
+    val progress  = new ProgressLogger(logger)
+
+    in.foreach { rec =>
+      totalRecords += 1
+
+      val umiAttribute = rec.getStringAttribute(umiTag)
+      if (umiAttribute == null || umiAttribute.isEmpty) {
+        if (missingUmisRecords == 0) logger.warning(s"Read (${rec.getReadName}) detected without UMI in tag $umiTag")
+        missingUmisRecords += 1
+        rejectOut.foreach(w => w.addAlignment(rec))
+      }
+      else {
+        val sequences = umiAttribute.split('-')
+        if (sequences.exists(_.length != umiLength)) {
+          if (wrongLengthRecords == 0) logger.warning(s"Read (${rec.getReadName}) detected with unexpected length UMI(s): ${sequences.mkString(" ")}")
+          wrongLengthRecords += 1
+          rejectOut.foreach(w => w.addAlignment(rec))
+        }
+        else {
+          // Find matches for all the UMIs
+          val matches = sequences.map(findBestMatch(_, umiSequences))
+
+          // Update the metrics
+          matches.foreach { m =>
+            if (m.matched) {
+              val metric = umiMetrics(m.umi)
+              metric.total_matches += 1
+              m.mismatches match {
+                case 0 => metric.perfect_matches      += 1
+                case 1 => metric.one_mismatch_matches += 1
+                case 2 => metric.two_mismatch_matches += 1
+                case _ => metric.other_matches        += 1
+              }
+            }
+            else {
+              umiMetrics(unmatchedUmi).total_matches += 1
+            }
+          }
+
+          // Output the corrected read
+          if (matches.forall(_.matched)) {
+            val correctedUmi = matches.map(_.umi).mkString("-")
+            rec.setAttribute(this.umiTag, correctedUmi)
+            out.addAlignment(rec)
+          }
+          else {
+            mismatchedRecords += 1
+            rejectOut.foreach(r => r.addAlignment(rec))
+          }
+        }
+      }
+
+      progress.record(rec)
+    }
+
+    in.safelyClose()
+    out.close()
+    rejectOut.foreach(_.close())
+
+    // Finalize the metrics
+    val sortedMetrics        = umiMetrics.values.toSeq.sortBy(_.umi)
+    val totalWithUnmatched   = sortedMetrics.map(_.total_matches).sum.toDouble
+    val meanWithoutUnmatched = sortedMetrics.filter(_.umi != unmatchedUmi).map(_.total_matches).sum / (sortedMetrics.size - 1d)
```

Comment:
It's a bit confusing, but `sortedMetrics` is the set of all UMIs + 1 entry for unmatched.  So min size is 2, and that's why the `-1` is there.